src/pattern_matching/myers.rs

```diff
@@ -156,29 +446,374 @@ impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
+        if num_cols > curr_len {
+            self.states.reserve(num_cols);
+            self.states.extend((0..num_cols - curr_len).map(|_| State::new(0)));
+        } else {
+            self.states.truncate(num_cols);
+        }
+        // important if using unsafe in add_state()
+        debug_assert!(self.states.len() == num_cols);
+
+        // first column is used to ensure a correct path if the text
+        // is shorter than the pattern
+        // Actual leftmost column starts at second position
+        let state0 = &mut self.states[0];
+        state0.dist = m;
+        state0.pv = T::max_value(); // all 1s
+        // the first position is 1, not 0 (after call to add_state)
+        self.pos = self.positions.next().unwrap();
+
+        self.m = m;
+    }
+
+    #[inline]
+    fn add_state(&mut self, s: State<T>) {
+        self.pos = self.positions.next().unwrap();
+        //self.states[self.pos] = s;
+        // faster and safe since the positions are cycled.
+        // Unless the code in init() is wrong, the index should
+        // never be out of bounds.
+        *unsafe { self.states.get_unchecked_mut(self.pos) } = s;
+    }
+
+    // Returns the length of the current match, optionally adding the
+    // alignment path to `ops`
+    fn traceback(&self, ops: Option<&mut Vec<AlignmentOperation>>) -> (usize, usize) {
+        self._traceback_at(self.pos, ops)
+    }
+
+    // Returns the length of a match with a given end position, optionally adding the
+    // alignment path to `ops`
+    // only to be called if the `states` vec contains all states of the text
+    fn traceback_at(&self, pos: usize, ops: Option<&mut Vec<AlignmentOperation>>) -> Option<(usize, usize)> {
+        let pos = pos + 1; // in order to be comparable since self.pos starts with 1, not 0
+        if pos <= self.pos {
+            return Some(self._traceback_at(pos, ops));
+        }
+        None
+    }
+
+    // returns a tuple of alignment length and hit distance, optionally adding the alignment path
+    // to `ops`
+    #[inline]
+    fn _traceback_at(
+        &self,
+        pos: usize,
+        mut ops: Option<&mut Vec<AlignmentOperation>>
+    )
+    -> (usize, usize)
+    {
+
+        use self::AlignmentOperation::*;
+
+        // Reverse iterator over states. If remembering all positions,
+        // the chain() and cycle() are not actually needed, but there seems
+        // to be almost no performance loss.
+        let mut states = self.states[..pos + 1].iter().rev().chain(
+            self.states.iter().rev().cycle()
+        );
+        // // Simpler alternative using skip() is slower in some cases:
+        // let mut states = self.states.iter().rev().cycle().skip(self.states.len() - pos - 1);
+
+        let ops = &mut ops;
+        if let Some(o) = ops.as_mut() {
+            o.clear();
+        }
+
+        // Mask with one bit switched on at the current position
+        let max_mask = T::one() << (self.m - 1);
```

Comment:
This should actually be like "Bit mask covering all positions of the pattern"

Reply:
Ok, please add that to the comment.